Implemented forgot pw function

miyabitanimchi opened the pull request:
I created a forgot pw function. 
To test, you can sign up with a your real email and put your email on `/forgotpassword` page,  and you'll receive en email which has a link expires in 15 mins. (we can change the limit anytime later)

There is a bug that the image on the left side on `/forgotpassword/reset/:id` page is not loaded for some reason, I'll figure out why later.

I'm going to refactor and put validation on login and signup on frontend later too, but please check the function when you have time, thank you!

Dannyboy222 reviewed (approved):
thank you

shiv-chan commented on client/src/features/forgotpassword/ForgotPassword.tsx:
```diff
@@ -1,37 +1,76 @@
+                <Link to="/forgotpassword/message">
+                  <button
+                    className="button w-full"
+                    onClick={(e) => sendResetPwEmail(e)}
```
```suggestion
                    onClick={sendResetPwEmail}
```
you can make it shorter like this.  The same goes for the rest.

shiv-chan commented on client/src/features/forgotpassword/ForgotPasswordReset.tsx:
```diff
@@ -1,101 +1,177 @@
 import { AiFillEye, AiFillEyeInvisible } from 'react-icons/ai';
+import ForgotPasswordInvalidUrl from './ForgotPasswordInvalidUrl';
+import axios from 'axios';
+const PASSWORD_REGEX =
+  /^(?=.*[a-z])(?=.*[0-9])[a-zA-Z0-9~`! @#\$%\^&*()_\-\+=\{\[\}\]\|\\:;"'<,>\.\?/]{6,32}$/;
```
@miyabitanimchi You could have used the `validator` npm package here as well, but this is also fine👍

shiv-chan commented on client/src/features/forgotpassword/ForgotPasswordReset.tsx:
```diff
@@ -1,101 +1,177 @@
-		</main>
-	);
+  useEffect(() => {
+    const verifyUrl = async () => {
+      try {
+        console.log('verifying');
+        const response = await axios.get(
+          `http://localhost:5000/forgotpassword/reset/${hasheduserid}`
+        );
+        console.log(response.data.message);
+        setIsUrlValid(true);
+      } catch (error: any) {
+        console.log(`the url is already expired! ${error}`);
+        if (error.response) {
+          console.error(error.response.data.message);
+        } else {
+          console.error(error);
+        }
+        setIsUrlValid(false);
+      }
+    };
```
**Suggestion:** place the function `verifyUrl` outside of `useEffect` or invoke this function immediately 💡

shiv-chan commented on client/src/features/forgotpassword/ForgotPasswordReset.tsx:
```diff
@@ -1,101 +1,177 @@
+    verifyUrl();
+  }, []);
+
+  useEffect(() => {
+    return;
+  }, [isVisible]);
```
Is this necessary?

miyabitanimchi replied:
Same as above, I'll remove it 🙏

shiv-chan commented on server/controllers/forgotPassword/updatePassword.js:
```diff
@@ -0,0 +1,48 @@
+    res.status(201).json({ message: 'password updated!' });
+    console.log('password updated!');
+  } catch (error) {
+    console.log(error);
```
We need to return any response from here so that the client-side catches an error.

shiv-chan commented on server/controllers/forgotPassword/updatePassword.js:
```diff
@@ -0,0 +1,48 @@
+      runValidatorOption
+    ).exec();
+    const forgotPasswordUser = await forgotPwUser.findOne({
+      userId: decodedUserId,
```
```suggestion
      userId: decodedUserId._id,
```
Is it okay to specify the key `_id`?
I believe that `decodedUerId` would be an object like `{ _id: <user id> }` when it's verified.
Please refer to https://www.npmjs.com/package/jsonwebtoken#jwtverifytoken-secretorpublickey-options-callback

shiv-chan reviewed (commented):
Thank you and great progress, @miyabitanimchi 🔥 

I checked how it looks and works on the browser.
When you try to access `forgotpassword/reset/:hashedid` with a random id, nothing will be shown. (paused with a white screen) Please check the comment that I left on `server/controllers/forgotPassword/verifyUrl.js`.

Also, there are some questions or suggestions, comments here.
Please check them and fix some as needed.

Thanks!

miyabitanimchi commented on client/src/common/components/LeftPart.tsx:
```diff
@@ -1,31 +1,38 @@
-	);
+  const location = useLocation();
+
+  useEffect(() => {
+    return;
+  }, [location.pathname]);
```
It wrote this because I thought earlier that the component wouldn't render again without the `useEffect()`, but I now confirmed that it's not needed, so I'm gonna remove it. Thank you for pointing it out!🙏

miyabitanimchi commented on client/src/features/forgotpassword/ForgotPassword.tsx:
```diff
@@ -1,37 +1,76 @@
+      setIsEmailsent(true);
+      // setEmail("");
+    } catch (error) {
+      console.log(`this is error ${error}`);
```
I already changed `log` to `error` on `feature-enhance-login-signup-forgotPw` branch, but thanks for your suggestion!

miyabitanimchi commented on client/src/features/forgotpassword/ForgotPassword.tsx:
```diff
@@ -1,37 +1,76 @@
+                </label>
+                <div className="relative">
+                  <input
+                    type="email"
```
Thank you💡Same as above, I added the id on each input on `feature-enhance-login-signup-forgotPw` 👍

miyabitanimchi commented on client/src/features/forgotpassword/ForgotPasswordMessage.tsx:
```diff
@@ -3,24 +3,24 @@ import LeftPart from '../../common/components/LeftPart';
+      <section className="flex justify-center my-auto h-9/12 w-1/2 tablet_l_max:w-full">
+        <div className="flex flex-col justify-center text-center h-full w-10/12 mobile_xl_max:w-9/12">
+          <p className="flex-initial text-2xl mb-2 mobile_l_max:text-xl">
+            Please check your email and a new password,
```
Again I added a verb "create" on `feature-enhance-login-signup-forgotPw` branch but I forgot to change , to .
I think I'll change it to "set" and add period on this branch
Thank you!

miyabitanimchi commented on client/src/features/forgotpassword/ForgotPasswordReset.tsx:
```diff
@@ -1,101 +1,177 @@
 import { AiFillEye, AiFillEyeInvisible } from 'react-icons/ai';
+import ForgotPasswordInvalidUrl from './ForgotPasswordInvalidUrl';
+import axios from 'axios';
+const PASSWORD_REGEX =
+  /^(?=.*[a-z])(?=.*[0-9])[a-zA-Z0-9~`! @#\$%\^&*()_\-\+=\{\[\}\]\|\\:;"'<,>\.\?/]{6,32}$/;
```
I added the validator npm on the client side on `feature-enhance-login-signup-forgotPw` branch👍

miyabitanimchi commented on client/src/features/forgotpassword/ForgotPasswordReset.tsx:
```diff
@@ -1,101 +1,177 @@
+                  <input
+                    type={isVisible ? 'text' : 'password'}
+                    className="h-10 mb-12 w-full px-8"
+                    onChange={(e) => handleOnChange(e)}
```
I actually didn't know before that both `{() => name()}` and `{name}` behave the same. Thank you for pointing this out and it was really good to know  🙈

miyabitanimchi commented on server/controllers/forgotPassword/updatePassword.js:
```diff
@@ -0,0 +1,48 @@
+    }
+    const saltPassword = await bcrypt.genSalt(10);
+    const hashedPassword = await bcrypt.hash(password, saltPassword);
+    const runValidatorOption = { runValidators: true };
```
No it didn't..., so I added another validator on the client and server side on `feature-enhance-login-signup-forgotPw` branch. I'm gonna remove the code on the line 31

miyabitanimchi commented on server/controllers/forgotPassword/verifyUrl.js:
```diff
@@ -0,0 +1,32 @@
+    });
+    console.log(forgotPasswordUser);
+    // 900000ms = 15min see how much time passed
+    const timePassed = forgotPasswordUser
+      ? timeAccessedToUrlAt - forgotPasswordUser.urlCreatedAt
+      : 900000;
```
Yeah, because there is the conditional 
` if (forgotPasswordUser && timePassed < 900000) { `
      `res.status(201).json({ message: 'url still valid!' });`
      `console.log('url still valid!');`
    `} else {`
right after the variable
it should be okay without the ternary operator. I'll double check later just in case but I'm going to remove it if it's not needed. Thank you!

miyabitanimchi commented on server/routes/forgotPw.js:
```diff
@@ -0,0 +1,18 @@
+import { verifyUrl } from '../controllers/forgotPassword/verifyUrl.js';
+const app = express();
+
+app.use('/reset/:hasheduserid', (req, res, next) => {
+  console.log('Request Id:', req.params.hasheduserid);
+  next();
+});
```
I checked the docs of Express again but maybe no? 🤔 Let me test it later and if I confirm we don't need the middleware let me remove the part 
<img width="705" alt="スクリーンショット 2021-12-11 午後5 17 26" src="https://user-images.githubusercontent.com/69213541/145696869-f28e6ad0-3f77-45bf-86a1-5ca6791192d0.png">

miyabitanimchi commented:
I improved code @shiv-chan pointed out with comments (As I mentioned for some parts I already changed on `feature-enhance-login-signup-forgotPw`, sorry for the inconvenience 🙏)
And I put the variable `decodedUserId ` on `server/controllers/forgotPassword/verifyUrl.js` in the `try` and `catch`, and also added the response so the client receive the error and show the error message on a browser.
(I added props on `client/src/features/forgotpassword/ForgotPasswordInvalidUrl.tsx`)

And we didn't need the middleware mentioned [here](https://github.com/shiv-chan/senpai-kohai/pull/20#discussion_r767210647). It still worked without this. 

Please review again when you have time, thank you!

shiv-chan commented on client/src/common/components/LeftPart.tsx:
```diff
@@ -1,32 +1,28 @@
-  useEffect(() => {
-    return;
-  }, [location.pathname]);
+  const { pathname } = useLocation();
```
It's nice and became simpler 👍

shiv-chan reviewed (approved):
Thanks, @miyabitanimchi for updating!
Looks good to me🙆‍♀️

shiv-chan commented:
@miyabitanimchi I'm still a bit confused with #20 and #21 because it was kinda difficult to see what is fixed in which...so let's check the points that I noticed or you updated are working okay after both #20 and #21 are merged to `develop`.
